normandy/recipes/api/views.py

```diff
@@ -119,42 +99,97 @@ def history(self, request, pk=None):
-                            status=status.HTTP_400_BAD_REQUEST)
-        recipe.approver = request.user
+        recipe.enable()
         recipe.save()
-
         return Response(status=status.HTTP_204_NO_CONTENT)
 
-    @transaction.atomic()
-    @reversion.create_revision()
+    @reversion_transaction
     @detail_route(methods=['POST'])
-    def enable(self, request, pk=None):
+    def disable(self, request, pk=None):
         recipe = self.get_object()
+        recipe.disable()
+        recipe.save()
+        return Response(status=status.HTTP_204_NO_CONTENT)
 
-        try:
-            recipe.enable()
-        except recipe.IsNotApproved:
-            return Response({'is_approved': 'This recipe cannot be enabled until it is approved.'},
+    @detail_route(methods=['GET'])
+    def approval_requests(self, request, pk=None):
+        recipe = self.get_object()
+        serializer = ApprovalRequestSerializer(recipe.approval_requests, many=True)
+        return Response(serializer.data)
+
+
+class ApprovalRequestViewSet(UpdateOrCreateModelViewSet):
+    queryset = ApprovalRequest.objects.all()
+    serializer_class = ApprovalRequestSerializer
+    filter_fields = ('active',)
+    permission_classes = [
+        permissions.DjangoModelPermissionsOrAnonReadOnly,
+        AdminEnabledOrReadOnly,
+    ]
+
+    def handle_exception(self, exc):
+        if isinstance(exc, ApprovalRequest.ActiveRequestAlreadyExists):
+            return Response(
+                {'active': 'You can only have one open approval request for a recipe.'},
+                status=status.HTTP_400_BAD_REQUEST)
+        elif isinstance(exc, ApprovalRequest.IsNotActive):
+            return Response({'active': 'This approval request has already been closed.'},
                             status=status.HTTP_400_BAD_REQUEST)
-        else:
-            return Response(status=status.HTTP_204_NO_CONTENT)
 
-    @transaction.atomic()
-    @reversion.create_revision()
+        return super().handle_exception(exc)
+
+    @reversion_transaction
+    def create(self, request, *args, **kwargs):
+        return super().create(request, *args, **kwargs)
+
+    @reversion_transaction
+    def update(self, request, *args, **kwargs):
+        return super().update(request, *args, **kwargs)
+
+    @reversion_transaction
     @detail_route(methods=['POST'])
-    def disable(self, request, pk=None):
-        recipe = self.get_object()
-        recipe.disable()
+    def approve(self, request, pk=None):
+        approval_request = self.get_object()
+        approval_request.approve(request.user)
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @detail_route(methods=['POST'])
+    def reject(self, request, pk=None):
+        approval_request = self.get_object()
+        approval_request.reject()
         return Response(status=status.HTTP_204_NO_CONTENT)
 
+    @detail_route(methods=['POST'])
+    def comment(self, request, pk=None):
+        if 'text' not in request.data:
```

Comment:
Did you try using the `ApprovalRequestComment` serializer for this validation? The docs say that DRF might even handle generating a response automatically if you use the `is_valid(raise_exception=True)` method on the serializer.